File: iosgl/iosgl/bt/display/particle/model/Display3DModelPartilce.m
//
//  Display3DModelPartilce.m
//  iosgl
//
//  Created by zhao on 15/4/2020.
//  Copyright © 2020 zhao. All rights reserved.
//

#import "Display3DModelPartilce.h"
#import "Context3D.h"
#import "TextureManager.h"
#import "DynamicTexItem.h"
#import "Scene3D.h"
#import "TextureRes.h"
#import "MaterialManager.h"
#import "ObjData.h"
#import "Scene_data.h"
#import "ParticleModelData.h"

@interface Display3DModelPartilce ()
@property (nonatomic, strong)  TextureRes* testTextureRes;
@property (nonatomic, assign)  BOOL  isTrue;
@end


@implementation Display3DModelPartilce

- (void)update;
{
    [super update];
}
 
 
- (void)setVc;
{
      [self setViewCamModeMatr3d];
   
      [self updateRotaionMatrix];
      Context3D *ctx=self.scene3d.context3D;
      [ctx setVcMatrix4fv:self.shader3d name:"rotMatrix" data:self.rotationMatrix3D.m];
}
- (void)setVa;
{
    Context3D *ctx=self.scene3d.context3D;
    ObjData* temp=self.facetdata.objData;
    
    [ctx pushVa: temp.verticesBuffer];
    [ctx setVaOffset:self.shader3d name:"v3Position" dataWidth:3 stride:0 offset:0];
    [ctx pushVa: temp.uvBuffer];
    [ctx setVaOffset:self.shader3d name:"v2TexCoord" dataWidth:2 stride:0 offset:0];
     
    [ctx drawCall:temp.indexBuffer  numTril:temp.trinum ];
    
}
- (void)resetVa;
{
   Context3D *ctx=self.scene3d.context3D;
    [ctx clearVa:0];
    [ctx clearVa:1];
}
 
-(void)updateRotaionMatrix;
{
    Camera3D *cam=self.scene3d.camera3D;
    Display3DModelPartilce* this=self;
    [this.rotationMatrix3D identity];
    if (this.data._watchEye) {
        //        this.timeline.inverAxisRotation(this._rotationMatrix);
        //        this._rotationMatrix.prependRotation(-Scene_data.cam3D.rotationY, Vector3D.Y_AXIS);
        //        this._rotationMatrix.prependRotation(-Scene_data.cam3D.rotationX, Vector3D.X_AXIS);
    }
    if (this.data._isZiZhuan) {
        [this.timeLine applySelfRotation:this.rotationMatrix3D axis:this.data._ziZhuanAngly];
    }
 
     
}
-(ParticleModelData*)facetdata;
{
    return (ParticleModelData*)self.data;
}
@end

File: test/three/app/src/main/java/com/z3d/display/particle/model/Display3DModelPartilce.java
package com.z3d.display.particle.model;

import com.z3d.base.Camera3D;
import com.z3d.base.ObjData;
import com.z3d.core.Context3D;
import com.z3d.display.particle.Display3DParticle;

public class Display3DModelPartilce  extends Display3DParticle {

    @Override
    public void update() {
        super.update();
    }
    @Override
    public void setVc() {
        super.setVc();
        Context3D ctx=this.scene3d.context3D;
        this.setViewCamModeMatr3d();
        this.updateRotaionMatrix();
        ctx.setVcMatrix4fv(this.shader3D,"rotMatrix",this.rotationMatrix3D.m);
    }
    @Override
    public void setVa() {
        super.setVa();
        Context3D ctx=this.scene3d.context3D;
        ObjData objData= this.modelData().objData;
        ctx.setVa(this.shader3D,"v3Position",3,objData.vertexBuffer);
        ctx.setVa(this.shader3D,"v2TexCoord",2,objData.uvBuffer);
        ctx.drawCall(objData.indexBuffer,objData.treNum);
    }
    private void updateRotaionMatrix() {
        Camera3D cam=this.scene3d.camera3D;
        this.rotationMatrix3D.identity();
        if (this.data._watchEye) {
        }
        if (this.data._isZiZhuan) {
            this.timeline.applySelfRotation(this.rotationMatrix3D,this.data._ziZhuanAngly);
        }
    }
    public ParticleModelData modelData(){
        return (ParticleModelData) this.data;
    }
}
